br/pkg/backup/prepare_snap/prepare.go

```diff
@@ -0,0 +1,420 @@
+	om.AddInt("inflight_requests", len(p.inflightReqs))
+	for _, r := range p.inflightReqs {
+		om.AddString("simple_inflight_region", region{maybeID: r.Id, startKey: r.StartKey, endKey: r.EndKey}.String())
+	}
+	om.AddInt("failed_requests", len(p.failedRegions))
+	for _, r := range p.failedRegions {
+		om.AddString("simple_failed_region", r.String())
+	}
+	err := om.AddArray("connected_stores", zapcore.ArrayMarshalerFunc(func(ae zapcore.ArrayEncoder) error {
+		for id := range p.clients {
+			ae.AppendUint64(id)
+		}
+		return nil
+	}))
+	if err != nil {
+		return err
+	}
+	om.AddInt("retry_time", p.retryTime)
+	om.AddBool("wait_apply_finished", p.waitApplyFinished)
+	return nil
+}
+
+// DriveLoopAndWaitPrepare drives the state machine and block the
+// current goroutine until we are safe to start taking snapshot.
+//
+// After this invoked, you shouldn't share this `Preparer` with any other goroutines.
+//
+// After this the cluster will enter the land between normal and taking snapshot.
+// This state will continue even this function returns, until `Finalize` invoked.
+// Splitting, ingesting and conf changing will all be blocked.
+func (p *Preparer) DriveLoopAndWaitPrepare(ctx context.Context) error {
+	logutil.CL(ctx).Info("Start drive the loop.", zap.Duration("retry_backoff", p.RetryBackoff),
+		zap.Int("retry_limit", p.RetryLimit),
+		zap.Duration("lease_duration", p.LeaseDuration))
+	p.retryTime = 0
+	if err := p.prepareConnections(ctx); err != nil {
+		log.Error("failed to prepare connections", logutil.ShortError(err))
+		return errors.Annotate(err, "failed to prepare connections")
+	}
+	if err := p.MaybeFinish(ctx); err != nil {
+		log.Error("failed to check the progress of our work", logutil.ShortError(err))
+		return errors.Annotate(err, "failed to begin step")
+	}
+	for !p.waitApplyFinished {
+		if err := p.WaitAndHandleNextEvent(ctx); err != nil {
+			log.Error("failed to wait and handle next event", logutil.ShortError(err))
+			return errors.Annotate(err, "failed to step")
+		}
+	}
+	return nil
+}
+
+// Finalize notify the cluster to go back to the normal mode.
+// This will return an error if the cluster has already entered the normal mode when this is called.
+func (p *Preparer) Finalize(ctx context.Context) error {
+	eg := new(errgroup.Group)
+	for id, cli := range p.clients {
+		cli := cli
+		id := id
+		eg.Go(func() error {
+			if err := cli.Finalize(ctx); err != nil {
+				return errors.Annotatef(err, "failed to finalize the prepare stream for %d", id)
+			}
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		logutil.CL(ctx).Warn("failed to finalize some prepare streams.", logutil.ShortError(err))
+		return err
+	}
+	logutil.CL(ctx).Info("all connections to store have shuted down.")
+	for {
+		select {
+		case event := <-p.eventChan:
+			if err := p.onEvent(ctx, event); err != nil {
+				return err
+			}
+		default:
+			return nil
+		}
+	}
+}
+
+func (p *Preparer) batchEvents(evts *[]event) {
+	for {
+		select {
```

Comment:
maybe also handle <-ctx.Done()

Reply:
In this context we won't block. I think we don't need to poll the context status here.